### PR TITLE
statistics for SQL/Mongo heuristics

### PR DESCRIPTION
**Pull request by jgaleotti**

### CODE REVIEW

File: client-java/sql/src/main/java/org/evomaster/client/java/sql/internal/SqlDistanceWithMetrics.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.evomaster.client.java.sql.internal;
+
+public class SqlDistanceWithMetrics {
+
+    public double sqlDistance;
+
+    public int numberOfEvaluatedRows;
+
+    public SqlDistanceWithMetrics(double sqlDistance, int numberOfEvaluatedRows) {
+        this.sqlDistance = sqlDistance;
```
Comment (arcuri82) [inline client-java/sql/src/main/java/org/evomaster/client/java/sql/internal/SqlDistanceWithMetrics.java:10]: should do some checks on input here. eg, throw a `IllegalArgumentException` if any of the inputs is smaller than 0

File: core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt [hunk 2]
```diff
@@ -63,6 +64,13 @@ class Statistics : SearchListener {
      */
     private var coverageFailures = 0
 
+    var sqlHeuristicsEvaluationCount = 0
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt:67]: why is this public?

File: core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt [hunk 3]
```diff
@@ -63,6 +64,13 @@ class Statistics : SearchListener {
...
+
+    private var sqlHeuristicsTotalNumberOfEvaluatedRows = 0
+
+    var mongoHeuristicsEvaluationCount = 0
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt:71]: why is this public?

File: core/src/test/kotlin/org/evomaster/core/search/service/StatisticsTest.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+class StatisticsTest {
+
+    @Test
+    fun testMongoHeuristicsAverage() {
+        val statistics = Statistics()
+        statistics.reportNumberOfEvaluatedDocumentsForMongoHeuristic(10)
+        statistics.reportNumberOfEvaluatedDocumentsForMongoHeuristic(20)
+        statistics.reportNumberOfEvaluatedDocumentsForMongoHeuristic(30)
+
+        assertEquals(3, statistics.mongoHeuristicsEvaluationCount)
```
Comment (arcuri82) [inline core/src/test/kotlin/org/evomaster/core/search/service/StatisticsTest.kt:15]: i see... still put those value as `private` and then declare a `public get`, eg something like:
```
private mongoHeuristicsEvaluationCount : Int
      public get
```

**Review by arcuri82 — CHANGES_REQUESTED** [review]

File: core/src/test/kotlin/org/evomaster/core/search/service/StatisticsTest.kt (hunk 4 above)
Comment (jgaleotti) [reply core/src/test/kotlin/org/evomaster/core/search/service/StatisticsTest.kt:15]: ok

File: core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt (hunk 3 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt:71]: modified to private

File: core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt (hunk 2 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/service/Statistics.kt:67]: modified to private

File: client-java/sql/src/main/java/org/evomaster/client/java/sql/internal/SqlDistanceWithMetrics.java (hunk 1 above)
Comment (jgaleotti) [reply client-java/sql/src/main/java/org/evomaster/client/java/sql/internal/SqlDistanceWithMetrics.java:10]: ok. I added these requirements to MongoDistanceWithMetrics too. I also added unit tests for this.

File: client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/db/MongoHandler.java [hunk 5]
```diff
@@ -84,66 +92,128 @@ public void setExtractMongoExecution(boolean extractMongoExecution) {
...
         for (Object doc : documents) {
-            double dist = calculator.computeExpression(info.getQuery(), doc);
-            if (dist == 0) return 0;
-            if (dist < min) min = dist;
+            numberOfEvaluatedDocuments += 1;
+            double findDistance;
+            try {
+                findDistance = calculator.computeExpression(info.getQuery(), doc);
+            } catch (Exception ex) {
+                findDistance = Double.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (jgaleotti) [reply client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/db/MongoHandler.java:161]: absolutely right. Thanks for pointing this out.

File: client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/db/MongoHandler.java [hunk 6]
```diff
@@ -84,66 +92,128 @@ public void setExtractMongoExecution(boolean extractMongoExecution) {
...
+        MongoExecutionsDto dto = new MongoExecutionsDto();
+        dto.failedQueries = emptyCollections.stream().map(this::extractRelevantInfo).collect(Collectors.toList());
         return dto;
     }
 
-    private double computeDistance(MongoInfo info) {
-        Iterable<?> documents = info.getDocuments();
+    private static Class<?> getCollectionClass(Object collection) throws ClassNotFoundException {
+        // collection is an implementation of interface MongoCollection
+        return collection.getClass().getInterfaces()[0];
```
[the remark this answers is not in the document]
Comment (jgaleotti) [reply client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/db/MongoHandler.java:126]: I agree.

**Review by arcuri82 — APPROVED** [review]

**Comment by arcuri82** [comment]
@jgaleotti approved. but waiting CI to finish before merging

**Pull request merged**